File: ASpinnerSearchFB/app/src/main/java/com/example/aspinnersearchfb/MainActivity.java
package com.example.aspinnersearchfb;

import android.support.annotation.NonNull;
import android.support.design.widget.BottomSheetDialog;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.example.aspinnersearchfb.Interface.IFirebaseLoadDone;
import com.example.aspinnersearchfb.Model.Movie;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.squareup.picasso.Picasso;
import com.toptoche.searchablespinnerlibrary.SearchableSpinner;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements IFirebaseLoadDone {

    SearchableSpinner searchableSpinner;

    DatabaseReference moviesRef;

    IFirebaseLoadDone iFirebaseLoadDone;

    List<Movie> movies;

    BottomSheetDialog bottomSheetDialog;

    //View
    TextView movie_title, movie_description;
    ImageView movie_image;
    FloatingActionButton btn_fav;

    boolean isFirstTimeClick = true;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        //Init Dialog
        bottomSheetDialog = new BottomSheetDialog(this);
        View bottom_sheet_dialog = getLayoutInflater().inflate(R.layout.layout_movie, null);

        movie_title = bottom_sheet_dialog.findViewById(R.id.movie_title);
        movie_description = bottom_sheet_dialog.findViewById(R.id.movie_description);
        movie_image = bottom_sheet_dialog.findViewById(R.id.movie_image);
        btn_fav = bottom_sheet_dialog.findViewById(R.id.btn_fav);

        //Event
        btn_fav.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //Close Dialog
                bottomSheetDialog.dismiss();
                Toast.makeText(MainActivity.this, "ADD FAV !!!", Toast.LENGTH_SHORT).show();
            }
        });

        //Set Content
        bottomSheetDialog.setContentView(bottom_sheet_dialog);

        searchableSpinner = findViewById(R.id.searchable_spinner);
        searchableSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                //Fixed First time click
                if (!isFirstTimeClick){
                    Movie movie = movies.get(position);
                    movie_title.setText(movie.getName());
                    movie_description.setText(movie.getDescription());
                    Picasso.get().load(movie.getImage()).into(movie_image);

                    bottomSheetDialog.show();
                }
                else {
                    isFirstTimeClick = false;
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        //Init DB
        moviesRef = FirebaseDatabase.getInstance().getReference("Movies");

        //Init Firebase
        iFirebaseLoadDone = this;

        //Get Data
        moviesRef.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                List<Movie> movies = new ArrayList<>();
                for (DataSnapshot movieSnapShot : dataSnapshot.getChildren()){
                    movies.add(movieSnapShot.getValue(Movie.class));
                }
                iFirebaseLoadDone.onFirebaseLoadSuccess(movies);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
                iFirebaseLoadDone.onFirebaseLoadFailed(databaseError.getMessage());
            }
        });


    }

    @Override
    public void onFirebaseLoadSuccess(List<Movie> movieList) {
        movies = movieList;

        //Get all name
        List<String> name_list = new ArrayList<>();
        for (Movie movie : movieList){
            name_list.add(movie.getName());
        }

        //Create Adapter and set for Spinner
        ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, name_list);
        searchableSpinner.setAdapter(adapter);
    }

    @Override
    public void onFirebaseLoadFailed(String message) {

    }
}

File: KotlinSpinnerSearchFB/app/src/main/java/com/example/kotlinspinnersearchfb/MainActivity.kt
package com.example.kotlinspinnersearchfb

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.design.widget.BottomSheetBehavior
import android.support.design.widget.BottomSheetDialog
import android.support.design.widget.FloatingActionButton
import android.view.View
import android.widget.*
import com.example.kotlinspinnersearchfb.Interface.IFirebaseLoadDone
import com.example.kotlinspinnersearchfb.Model.Movie
import com.google.firebase.database.*
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.activity_main.*
import java.util.ArrayList

class MainActivity : AppCompatActivity(), IFirebaseLoadDone {

    lateinit var moviesRef: DatabaseReference

    lateinit var iFirebaseLoadDone: IFirebaseLoadDone

    //Variable
    lateinit var bottomSheetDialog: BottomSheetDialog

    //View
    lateinit var movie_title: TextView
    lateinit var movie_description:TextView
    lateinit var movie_image: ImageView
    lateinit var btn_fav: FloatingActionButton

    internal var isFirstTimeClick = true

    internal var movies: List<Movie> = ArrayList<Movie>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //Init bottom sheet
        bottomSheetDialog = BottomSheetDialog(this)
        val bottom_sheet_dialog = layoutInflater.inflate(R.layout.layout_movie, null)

        movie_title = bottom_sheet_dialog.findViewById(R.id.movie_title)
        movie_description = bottom_sheet_dialog.findViewById(R.id.movie_description)
        movie_image = bottom_sheet_dialog.findViewById(R.id.movie_image)
        btn_fav = bottom_sheet_dialog.findViewById(R.id.btn_fav)

        //Event
        btn_fav.setOnClickListener {
            //Close Dialog
            bottomSheetDialog.dismiss()
            Toast.makeText(this@MainActivity, "ADD FAV !!!", Toast.LENGTH_SHORT).show()
        }

        //Set Content
        bottomSheetDialog.setContentView(bottom_sheet_dialog)

        searchable_spinner.setOnItemSelectedListener(object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                //Fixed First time click
                if (!isFirstTimeClick) {
                    val movie = movies.get(position)
                    movie_title.text = movie.name
                    movie_description.text = movie.description
                    Picasso.get().load(movie.image).into(movie_image)

                    bottomSheetDialog.show()
                } else {
                    isFirstTimeClick = false
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>) {

            }
        })

        //Init Firebase
        iFirebaseLoadDone = this

        //Init DB
        moviesRef = FirebaseDatabase.getInstance().getReference("Movies")

        //Load Data
        moviesRef.addValueEventListener(object : ValueEventListener {

            val movieList:MutableList<Movie> = ArrayList<Movie>()
            override fun onCancelled(p0: DatabaseError) {
                iFirebaseLoadDone.onFirebaseLoadFailed(p0.message)
            }

            override fun onDataChange(p0: DataSnapshot) {
                for (movieSnapShot in p0.children) {
                    movieList.add(movieSnapShot.getValue<Movie>(Movie::class.java)!!)
                }
                iFirebaseLoadDone.onFirebaseLoadSuccess(movieList)
            }
        })
    }

    override fun onFirebaseLoadSuccess(movieList: List<Movie>) {

        this.movies = movieList

        //Get all name from list
        val movie_name_title = getMovieNameList(movieList)

        //Create Adapter and set for Spinner
        val adapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, movie_name_title)
        searchable_spinner.adapter = adapter
    }

    private fun getMovieNameList(movieList: List<Movie>): List<String> {
        val result = ArrayList<String>()
        for (movie in movieList){
            result.add(movie.name!!)
        }
        return result
    }

    override fun onFirebaseLoadFailed(message: String) {

    }
}
